### PR TITLE
Avoid unnecessary copies & allocations in field selectors

### PR DESCRIPTION
**Pull request by wojtek-t**
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31248)

### CODE REVIEW

File: pkg/registry/event/strategy.go [hunk 1]
```diff
@@ -71,31 +70,31 @@ func (eventStrategy) AllowUnconditionalUpdate() bool {
...
-		"involvedObject.uid":             string(event.InvolvedObject.UID),
-		"involvedObject.apiVersion":      event.InvolvedObject.APIVersion,
-		"involvedObject.resourceVersion": event.InvolvedObject.ResourceVersion,
-		"involvedObject.fieldPath":       event.InvolvedObject.FieldPath,
-		"reason":                         event.Reason,
-		"source":                         event.Source.Component,
-		"type":                           event.Type,
-	}
-	return l, generic.MergeFieldsSets(objectMetaFieldsSet, specificFieldsSet), nil
+func SelectableFields(event *api.Event) fields.Set {
```
Comment (timothysc) [inline pkg/registry/event/strategy.go]: Should probably document the exposed f(n)

File: pkg/registry/event/strategy.go [hunk 2]
```diff
@@ -71,31 +70,31 @@ func (eventStrategy) AllowUnconditionalUpdate() bool {
...
+	fields := make(fields.Set, 12)
```
Comment (timothysc) [inline pkg/registry/event/strategy.go]: For my own edification, what is the advantage of the make over the implicit declaration?

**Comment by timothysc** [comment]
generally lgtm, but curious to know what the overall gain is?

File: pkg/registry/controller/strategy.go [hunk 3]
```diff
@@ -103,11 +103,10 @@ func (rcStrategy) AllowUnconditionalUpdate() bool {
...
 // ControllerToSelectableFields returns a field set that represents the object.
 func ControllerToSelectableFields(controller *api.ReplicationController) fields.Set {
-	objectMetaFieldsSet := generic.ObjectMetaFieldsSet(controller.ObjectMeta, true)
-	controllerSpecificFieldsSet := fields.Set{
-		"status.replicas": strconv.Itoa(int(controller.Status.Replicas)),
-	}
-	return generic.MergeFieldsSets(objectMetaFieldsSet, controllerSpecificFieldsSet)
+	fields := make(fields.Set, 3)
+	fields["status.replicas"] = strconv.Itoa(int(controller.Status.Replicas))
+	generic.AddObjectMetaFields(&controller.ObjectMeta, true, &fields)
```
Comment (lavalamp) [inline pkg/registry/controller/strategy.go]: if AddObjectMetaFields starts setting more than two fields, then the make() call should be changed. How will humans remember to do that?

File: pkg/registry/event/strategy.go (hunk 2 above)
Comment (lavalamp) [inline pkg/registry/event/strategy.go]: the parameter says how many elements to allocate space for, for optimizing allocations.

Comment (lavalamp) [inline pkg/registry/event/strategy.go]: I guess you're doing this because you know AddObjectMetaFields is going to add two more elements?

Comment (lavalamp) [inline pkg/registry/event/strategy.go]: Perhaps it'd be better to just explicitly add the elements from that here, and then have a test that it includes everything that AddObjectMetaFields would have added.

File: pkg/registry/namespace/strategy.go [hunk 4]
```diff
@@ -151,11 +151,10 @@ func MatchNamespace(label labels.Selector, field fields.Selector) *generic.Selec
...
-	specificFieldsSet := fields.Set{
-		"status.phase": string(namespace.Status.Phase),
-		// This is a bug, but we need to support it for backward compatibility.
-		"name": namespace.Name,
-	}
-	return generic.MergeFieldsSets(objectMetaFieldsSet, specificFieldsSet)
+	fields := make(fields.Set, 3)
+	fields["status.phase"] = string(namespace.Status.Phase)
+	// This is a bug, but we need to support it for backward compatibility.
+	fields["name"] = namespace.Name
```
Comment (lavalamp) [inline pkg/registry/namespace/strategy.go]: hm, there should be a field filter conversion func that handles this. That was the intended place to fix compatibility things.

File: pkg/registry/persistentvolume/strategy.go [hunk 5]
```diff
@@ -111,10 +111,9 @@ func MatchPersistentVolumes(label labels.Selector, field fields.Selector) *gener
 
 // PersistentVolumeToSelectableFields returns a field set that represents the object
 func PersistentVolumeToSelectableFields(persistentvolume *api.PersistentVolume) fields.Set {
-	objectMetaFieldsSet := generic.ObjectMetaFieldsSet(persistentvolume.ObjectMeta, false)
-	specificFieldsSet := fields.Set{
-		// This is a bug, but we need to support it for backward compatibility.
-		"name": persistentvolume.Name,
-	}
-	return generic.MergeFieldsSets(objectMetaFieldsSet, specificFieldsSet)
+	fields := make(fields.Set, 2)
```
Comment (lavalamp) [inline pkg/registry/persistentvolume/strategy.go]: 3?

**Comment by lavalamp** [comment]
is this a big effect? Seems to make the code slightly uglier :)

**Comment by wojtek-t** [comment]
>
is this a big effect? Seems to make the code slightly uglier :)

I initially thought that the biggest effect was from allocations to map, but it turned out, that the biggest effect is from avoiding copying ObjectMeta. So I can revert parts of this PR if you think this is uglier (personally I don't agree with that statement, but well..)

File: pkg/registry/controller/strategy.go (hunk 3 above)
Comment (wojtek-t) [inline pkg/registry/controller/strategy.go]: Obsolete (it seems that the real gains come from not copying ObjectMeta, so I'm reverting most of the changes in this PR.

**Comment by wojtek-t** [comment]
@lavalamp - this PR is 90%+ of the gains, and hopefully is not at all controversial.

PTAL

**Comment by wojtek-t** [comment]
@k8s-bot unit test this, issue: #31262

**Comment by lavalamp** [comment]
LGTM, thanks!

**Pull request merged**